### ISSUE TITLE
[Due for payment 2025-02-18] [Due for payment 2025-02-13] [$500] [Auth Violations] Support isDismissed on newDot to hide violations

### ISSUE THREAD

**Issue by pecanoro** [Expensify/App#53398]
# Part of the Auth Violations project
Main issue: https://github.com/Expensify/Expensify/issues/422533
Project: https://github.com/Expensify/Expensify/issues/393838

# Feature Description
Currently we have four violations that can be dismissed in old Expensify:

- DUPLICATED_TRANSACTION
- SMARTSCAN_FAILED
- RTER
- AUTOREPORTED_REJECTED_EXPENSE (Not implemented in newDot yet)

If the violation has been dismissed, we don't show the violation to the user. In PHP, we currently have a function like this:

```php
    /**
     * Checks whether a violation has been dismissed, optionally by a particular email address.
     *
     * @param string $violationName check Policy_Violation::NAME_*
     */
    public function isViolationDismissed(string $violationName, string $email = ''): bool
    {
        // If we don't provide an email, we just check if the violation has been dismissed by anyone
        if (!$email) {
            return count($this->getNVP('dismissedViolations')[$violationName] ?? []) > 0;
        }

        $timestamp = $this->getNVP('dismissedViolations')[$violationName][$email] ?? '';
        return strlen(strval($timestamp)) > 0;
    }
```

Note: The NVP is stored in the transaction's "comment" along with the rest of the name-value pairs.

Ideally, we should add this logic in newDot similar to what we did for duplicate transactions [here](https://github.com/Expensify/App/blob/3f375387b1881fe524e8b80cc3d1af4a456466de/src/libs/TransactionUtils/index.ts#L828-L830). My best guess is that we need to make that logic work in these [functions](https://github.com/Expensify/App/blob/3f375387b1881fe524e8b80cc3d1af4a456466de/src/libs/TransactionUtils/index.ts#L855-L889) here so we stop showing the violation in the LHN and the transaction itself. However, I might be missing other spots.

Take into account that checking for dismissed violations should be optional in case there are flows in which we want to show it regardless or block submission because of that.

# Manual Test Steps

1. Scan a receipt in which the smartscan will fail (illegible or not a receipt)
2. Go to old Expensify, find the report and dismiss the violation.
3. Go back to new Expensify and make sure the violation is not showing anymore. 
_Note: You might need to refresh the page, I am not sure if the BE is queueing these updates, let me know if not._ 

cc/ @cead22 @iwiznia 
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021864348688314983900</li>
        <li>Upwork Job ID: 1864348688314983900</li>
        <li>Last Price Increase: 2024-12-11</li>
<li>Automatic offers: </li>
<ul>
<li>rojiphil | Reviewer | 105371307</li>
<li>Krishna2323 | Contributor | 105371309</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @johncschuster</details>

**Comment by rojiphil** [comment]
@pecanoro, While testing the scenarios for resolving DUPLICATED_TRANSACTION and SMARTSCAN_FAILED from OD, it was noticed that the pusher updates had not been sent in ND. Refreshing the page though makes it work. So, we will need these to be pushed from BE. I have not yet checked RTER which I will do tomorrow my day and update. 

Meanwhile, where can we find the details related to  `AUTOREPORTED_REJECTED_EXPENSE`? If this is within a design document, please grant me access. Thanks.

**Comment by pecanoro** [comment]
> Meanwhile, where can we find the details related to AUTOREPORTED_REJECTED_EXPENSE? If this is within a design document, please grant me access. Thanks.

I don't think we have talked about implementing this yet but we should still implement this in a way that will work with all types of violations regardless.

**Comment by pecanoro** [comment]
What I mean is that even though we cannot test AUTOREPORTED_REJECTED_EXPENSE, the solution should cover any possible dismissed violation.

**Comment by dubielzyk-expensify** [comment]
Any visual changes as part of this change you want me to review from a design side?

**Comment by rojiphil** [comment]
> can you please review the proposals above?

@johncschuster Yeah... I was working on this. Will send my comments in a while.

**Comment by rojiphil** [comment]
Thanks all for your proposals.

> I thing it would be the best to use getTransactionViolations when getting violations for a specific transaction. This way we can fix all instances without adding the filter everywhere.

I like the idea of filtering the dismissedViolations in `getTransactionViolations` which can help us to have a generic approach without having to worry about individual cases 
@Krishna2323 Please note that we do not have to add `AUTOREPORTED_REJECTED_EXPENSE` to violations object as we would anyway have a generic solution.   

@pecanoro I think we can go with @Krishna2323 [proposal](https://github.com/Expensify/App/issues/53398#issuecomment-2538837692) which looks LGTM in general. More specifics about handling on the individual cases can be handled at the PR stage. 
🎀👀🎀 C+ reviewed

**Comment by johncschuster** [comment]
Heads up, I will be OOO from Monday, December 23, and returning Monday, January 6. I am not reassigning my assigned issues. If you feel this issue needs to be actioned by a Bug Zero team member, please reach out for help from any available team members in #expensify-open-source.

If not, I'll action this as soon as I'm back in office!

**Comment by pecanoro** [comment]
> @pecanoro after we dismiss the smartscan fail violation from OldDot and enter the details manually, a new warning type violation is received receiptNotSmartScanned. Is this expected?

Yeah, I think it only happens with that one, it's a weird one. They are different types of violations.

**Comment by rojiphil** [comment]
> do I need to record all the listed violations, or is recording only `SMARTSCAN_FAILED` sufficient?
> 
> 1. DUPLICATED_TRANSACTION
> 2. SMARTSCAN_FAILED
> 3. RTER

@Krishna2323 Did you mean testing the mentioned violations? If so, I think yes. Did you face any problems with this?

**Comment by pecanoro** [comment]
Sorry guys, I was on vacation for the holidays and I just returned. Let me check the last comment

**Comment by pecanoro** [comment]
> do I need to record all the listed violations, or is recording only SMARTSCAN_FAILED sufficient?

Ideally yes, you should. `DUPLICATED_TRANSACTION` and `SMARTSCAN_FAILED` are pretty straightforward. Is testing `RTER` too complicated?

**Comment by pecanoro** [comment]
Ahh got it! Just record all of them for a single platform; there is no need to record videos for all of them on all platforms. But make sure to test and record at least one on the other platforms.

**Comment by rojiphil** [comment]
@Krishna2323 Let us raise a new PR along with the optimizations implemented recently as the original PR is reverted. Can we optimize further by supporting `transactionIDs` in `useTransactionViolations`? And maybe, even optimize for [this](https://github.com/Expensify/App/pull/56405#discussion_r1945185712). We can discuss more in the PR itself.

**Comment by johncschuster** [comment]
@rojiphil can you please complete the BZ Checklist above? Once completed, I can issue payment.

**Comment by rojiphil** [comment]
# BugZero Checklist:

- [x] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [x] 1z. Other: New Feature

Where bug was reported:
  - [ ] 2a. Reported on production
  - [ ] 2b. Reported on staging (deploy blocker)
  - [ ] 2c. Reported on a PR
  - [x] 2z. Other: New Feature

Who reported the bug:
  - [ ] 3a. Expensify user
  - [x] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [x] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment: No offending PR as this was a new feature/improvement. 

- [x] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion: Not Required. 

- [x] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

## Regression Test Proposal
### Precondition:


### Test:
1. Scan a receipt in which the smartscan will fail (illegible or not a receipt)
2. Go to old Expensify, find the report and dismiss the violation.
3. Go back to new Expensify and verify that the violation is not shown anymore (Test 1) Note: You may have to refresh the page.
4. Create 2 manual expenses with same amount and merchant. This will result in `duplicate` transactions.
5. Go to old Expensify, find the report and dismiss the duplicate violation.
6. Go back to new Expensify and verify that the duplicate violation is not shown anymore. (Test 2) Note: You may have to refresh the page.

Do we agree 👍 or 👎

**Comment by rojiphil** [comment]
@johncschuster BZ Checklist done for adding test case to the test case repository. Otherwise, the checklist was not required as this was a new feature/improvement, not a bug. Thanks.

**Comment by johncschuster** [comment]
## Payment Summary

Contributor: @Krishna2323  paid $500 via [Upwork](https://www.upwork.com/nx/wm/offer/105371309)
Contributor+: @rojiphil owed $500 via NewDot

**Comment by rojiphil** [comment]
Requested in ND

**Comment by garrettmknight** [comment]
$500 approved for @rojiphil
